man/man1/uns2uns.1
.TH uns2uns 1NEMO "March 17, 2020"
.SH NAME
uns2uns \- a universal converting tool for UNS snapshots.

.SH SYNOPSIS
\fBuns2uns in=\fPuns_input\fB out=\fPuns_output\fB
select=\fPselection\fB type=\fPuns_type\fB [parameter=value] .\|.\|.
.SH DESCRIPTION
\fIuns2uns\fP reads an UNS N-body systems and outputs another UNS
N-body systems. It can convert an uns input file to an uns output
file. It can also copy out some components and/or physical quantities from
an uns file to another one.
.PP
In and Out can be in a different type (ex conversion from
Nemo to Gadget OR Gadget to Nemo). 
.PP
In and Out can also be in the same format and the user
can select a subset of the components and/or a subset of physical
quantities to save.
.PP
There are two kinds of nbody file structure:
.TP 0
- \fBrange\fP file structure (\fIRFS\fP), like NEMO, where  data are identified
  by range of indexes (ex 0:999 for the disk and 1000:9999  for the halo).  
.TP 0
- \fBcomponent\fP file structure (\fICFS\fP), like GADGET or RAMSES where data are identified by
  component name (disk, gas, halo, stars, bulge, bndry).
.SH PARAMETERS
The following parameters are recognized.
.TP 12
\fBin=\fP\fIuns input file\fP
Currently, the following input files are supported: Nemo, Gadget3,
Gadget2, Gadget1, RAMSES, list of files, sqlite3 uns data base
.TP
\fBout=\fP\fIuns output file\fP
Output data will be written to \fIout_file\fP in the format type
specified from the parameter \fBtype=\fP
.TP
\fBselect=\fP\fIinput/output particles selection\fP
this parameter select the input particles which will be saved in the
output file. There are several possible syntax. It can be a range of
index (RFS input, ex: select=0:999,3000:4000 ) or a list of components
(CFS input, ex: select=disk,stars), separated by coma.

There are a special syntax when in=RFS and out=CFS (ex Nemo to Gadget
conversion), which is : select=comp1@range+comp2@range (ex:
select=gas@0:1000+disk@300:400). For more details, see EXAMPLES below.
.TP
\fBtype=\fP\fIuns type format\fP
we specify here the type of the output file.
Currently, the following output files are supported: nemo, gadget2, gadget3
.TP
\fBbits=\fP\fIphysical quantities to saved\fP
If this parameter is left blank, then all physical quantities will be
saved, this is the default behaviour. Here is the list of differents
values:
  'm'  mass
  'x'  position
  'v'  velolcity
  'p'  potential
  'a'  acceleration
  'e'  softening
  'k'  keys array (from nemo)
  'X'  aux array  (from nemo)
  'R'  density
  'I'  particles's Id
  'U'  Internal energy
  'M'  metallicity
  'A'  stars's age
  'H'  hsml
  'T'  temperature

  ex: bits=xRH will save positions, densities and hsml
  
  (see also : 
https://projets.lam.fr/projects/unsio/wiki/UnsioVariablesUtilisation#How-to-select-data-at-loading)

.SH EXAMPLES
In the following examples, we assume that \fIfileg2\fP is a GADGET2 file,
\fIfilenemo\fP is a NEMO file, \fIfilelist\fP is a list of file and
\fIsimname\fP is a simulation's name stored in an uns sqlite3 database.
.TP 5
\fB- Gadget to Nemo conversion:\fP
1) uns2uns in=fileg2 out=filenemo select=disk,stars type=nemo bits=mx

Convert a gadget file to nemo file, by selecting the mass and the
positions of disk and stars components.

2) uns2uns in=filelist out=filenemo select=disk,stars type=nemo
bits=mx first=t

Do exactly the same conversion as 1) but from a list of file. Each
output files will have a trailing number.
.TP
\fB- Nemo to Gadget conversion:\fP
uns2uns in=filenemo out=fileg2 select=gas@0:100+halo@200:300 type=gadget2

Convert a Nemo file to Gadget2 by setting nemo particles from range
[0:100] to gadget2 gas component, and nemo particles from range
[200:300] to gadget2 halo component.
.TP
\fB- Physical quantities selection:\fP
1) uns2uns in=fileg2 out=fileg22 select=gas type=gadget2 bits=xRH

Create a gadget2 file from another gadget2 file, by keeping only
positions, densities and hsml of gas component.

2) uns2uns in=simname out=fileg22 select=gas type=gadget2 bits=xRH first=t

Do exactly the same thing as 1) but on all files stored in the uns
sqlite3 database which match to entry 'simname'. Each
output files will have a trailing number.

.SH SEE ALSO
snapmask(1NEMO), snapcopy(1NEMO), snapshot(5NEMO), uns_info(1NEMO)
.TP
https://projets.lam.fr/projects/unsio/wiki

.SH AUTHOR
Jean-Charles LAMBERT
.SH HISTORY
.nf
.ta +1i +4i
05-Jul-11	V1.0 created	JCL
17-Mar-20	V2.0 bugs fixed	JCL
.fi



usr/jcl/uns_project/src/uns2uns.cc
// ============================================================================
// Copyright Jean-Charles LAMBERT - 2011                                       
// e-mail:   Jean-Charles.Lambert@oamp.fr                                      
// address:  Dynamique des galaxies                                            
//           Laboratoire d'Astrophysique de Marseille                          
//           Pole de l'Etoile, site de Chateau-Gombert                         
//           38, rue Frederic Joliot-Curie                                     
//           13388 Marseille cedex 13 France                                   
//           CNRS U.M.R 6110                                                   
// ============================================================================
#include <iostream>                                   // C++ I/O     
#include <fstream>                                    // C++ file I/O
#include <sstream>
#include <cstdio>                   
#include <cstdlib>                  
#include <assert.h>
#include <cstdio>
#include <iomanip>
#include <nemo.h>       
#include <vector>
#include <map>

// ------------------------------------------------------------
// Include file
#include "uns.h"
#include "csnaptools.h"
using namespace std; 
using namespace jclut;

// NEMO parameters
const char * defv[] = {  // use `::'string because of 'using namespace std'
  "in=???\n           uns input file ",
  "out=???\n          uns output file (gagdet|nemo)",
  "select=???\n       select input particles to be saved (range or component)",
  "type=???\n         type of the outfile (gadget1|gadget2|nemo)",
  "bits=\n            physicals quantities that you want to save\n",
  "times=all\n         selected time                   ",
  "first=f\n           add a trailing numbering to the first output file",
  "offset=0.01\n      +/- time offset",
  "verbose=f\n        verbose mode                    "
  "VERSION=1.0\n       compiled on <"__DATE__"> JCL   ",
  NULL
};
const char * usage="A universal converting tool for UNS snapshots";
//------------------------------------------------------------------------------
// saveComponent
// save components according to their indexes stored in a vector
template <class T> void saveComponent(T * array, std::string typearray,
                                      const int n, 
                                      std::vector<int> &vec, 
                                      const int dim, 
                                      std::string comp, 
                                      uns::CunsOut * unsout)
{
  float * array_o = new float[vec.size()*dim]; // particles to be saved
  // loop on vector index to select particles to be save
  for (unsigned int i=0;i<vec.size();i++) {
    assert(vec[i]<n);
    for (int j=0;j<dim;j++) {
      array_o[i*dim+j] = array[vec[i]*dim+j];
    }
  }
  unsout->snapshot->setData(comp,typearray,vec.size(),array_o,false);
  delete [] array_o;
}
//------------------------------------------------------------------------------
// processFromRangeComp
// we parse range of particles from component parameters
// this method should be used with RANGE like input file, and COMPONENT output
// [select=] input string must be built like this
// ex : select=gas@0:10000+halo@200000:3000000
//  
// ALL particles MUST have been selected during CUNSIN instantiation
// read pos,vel,mass of the components
// if component exist AND it has been selected, then respecting comp's data are
// prepared to be saved
void processFromRangeComp(std::string select, uns::CunsIn * unsin,uns::CunsOut * unsout)
{
  float * pos, * vel, * mass;
  int n,nbody;
  bool ok;
  // parse select string which MUST have '@' separator
  if (select.find("@",0) == std::string::npos) { // "@" ! found
    std::cerr 
        << "\n\nprocessFromRangeComp : algorithm ERROR !!\n"
        << "you want to convert a RANGE to COMPONENT file format, but\n"
        << "your parameter [select="<<select<<"] is misformed and should be like:\n"
        << "example: select=gas@0:100+disk@200:300\n"
        << "aborting...\n\n";
    std::exit(1);
  }
  // get nbody in
  ok =unsin->snapshot->getData("nsel",&nbody);  
  // read range of particles for the component
  // and put then in a vector
  std::map <std::string, std::vector<int> > myMapOfVec;
  // build a map<component,vector of indexes>
  myMapOfVec = CSnaptools::mapStringVectorIndexes(select,nbody);
  // loop on each component of the map
  for( std::map <std::string, std::vector<int> >::iterator iter = myMapOfVec.begin(); iter != myMapOfVec.end(); ++iter ) {    
    std::string      comp = (*iter).first;   // component
    std::vector<int> vec  = (*iter).second;  // vector of indexes
    std::cerr << "["<<comp<<"] vec size = " << vec.size() << "\n";
    for (unsigned int i=0; i<vec.size();i++) {
      //std::cerr << "i="<<i<<"        -> " << vec[i] << "\n";
    }    
    // read position
    ok = unsin->snapshot->getData("all","pos" ,&n,&pos );
    if (ok) {
      saveComponent(pos,"pos",n,vec,3,comp,unsout);      
    }
    // read velocities
    ok = unsin->snapshot->getData("all","vel" ,&n,&vel );
    if (ok) {
      saveComponent(vel,"vel",n,vec,3,comp,unsout);      
    }
    // read masses
    ok = unsin->snapshot->getData("all","mass" ,&n,&mass );
    if (ok) {
      saveComponent(mass,"mass",n,vec,1,comp,unsout);      
    }
    // read Ids
    int * id;
    ok = unsin->snapshot->getData("all","id" ,&n,&id );
    if (ok) {
      saveComponent(id,"id",n,vec,1,comp,unsout);
    }
    // read accelerations
    float * acc;
    ok = unsin->snapshot->getData("all","acc" ,&n,&acc );
    if (ok) {
      saveComponent(acc,"acc",n,vec,3,comp,unsout);
    }
    // read pot
    float * pot;
    ok = unsin->snapshot->getData("all","pot" ,&n,&pot );
    if (ok) {
      saveComponent(pot,"pot",n,vec,1,comp,unsout);
    }
    if (comp=="gas") {
      float * rho, * hsml;
      // Try to get Rho
      ok = unsin->snapshot->getData("all","rho" ,&n,&rho );
      if (ok) {
        saveComponent(rho,"rho",n,vec,1,comp,unsout);
      }
      // Try to get Hsml
      ok = unsin->snapshot->getData("all","hsml" ,&n,&hsml );
      if (ok)
        saveComponent(hsml,"hsml",n,vec,1,comp,unsout);
    }
  }
}
//------------------------------------------------------------------------------
// processFromComp
// try to read all data of the components
// if component exist AND it has been selected, then respecting comp's data are
// prepared to be saved
void processFromComp(std::string comp, uns::CunsIn * unsin,uns::CunsOut * unsout)
{  
  float * pos, * vel, * mass;
  int nbody,n1,n2,n3;
  bool ok;
  // read position
  ok = unsin->snapshot->getData(comp,"pos" ,&n1,&pos );
  if (ok) {
    if (comp=="all") unsout->snapshot->setNbody(n1);
    unsout->snapshot->setData(comp,"pos",n1,pos,false);
  }
  // read velocities
  ok = unsin->snapshot->getData(comp,"vel" ,&n2,&vel );
  if (ok) {
    if (comp=="all") unsout->snapshot->setNbody(n2);
    unsout->snapshot->setData(comp,"vel",n2,vel,false);
  }
  // read masses
  ok = unsin->snapshot->getData(comp,"mass" ,&n3,&mass );
  if (ok) {
    if (comp=="all") unsout->snapshot->setNbody(n3);
    unsout->snapshot->setData(comp,"mass",n3,mass,false);
  }
  // read Ids
  int * id;
  ok = unsin->snapshot->getData(comp,"id" ,&nbody,&id );
  if (ok) { 
    if (comp=="all") unsout->snapshot->setNbody(nbody);
    unsout->snapshot->setData(comp,"id",nbody,id,false);
  }
  // read Pot
  float * pot;
  ok = unsin->snapshot->getData(comp,"pot" ,&nbody,&pot );
  if (ok) {
    if (comp=="all") unsout->snapshot->setNbody(nbody);
    unsout->snapshot->setData(comp,"pot",nbody,pot,false);
  }
  // read accelerations
  float * acc;
  ok = unsin->snapshot->getData(comp,"acc" ,&nbody,&acc );
  if (ok) {
    if (comp=="all") unsout->snapshot->setNbody(nbody);
    unsout->snapshot->setData(comp,"acc",nbody,acc,false);
  }
  if (comp=="all") {
    float * rho, * hsml;
    int nn;
    nbody=std::max(n1,std::max(n2,n3));
    // Try to get Rho
    ok = unsin->snapshot->getData("rho" ,&nn,&rho );
    if (ok && nbody == nn) {
      if (comp=="all") unsout->snapshot->setNbody(nn);
      unsout->snapshot->setData(comp,"rho",nbody,rho,false);
    }  
    // Try to get Hsml
    ok = unsin->snapshot->getData("hsml" ,&nn,&hsml );
    if (ok && nbody == nn)  {
      if (comp=="all") unsout->snapshot->setNbody(nn);
      unsout->snapshot->setData(comp,"hsml",nbody,hsml,false);  
    }
  }
  if (comp=="gas") {
    float * rho, * hsml, * u , * temp, * metal;
    // Try to get Rho
    ok = unsin->snapshot->getData(comp,"rho" ,&nbody,&rho );
    if (ok) {
      unsout->snapshot->setData(comp,"rho",nbody,rho,false);
    }  
    // Try to get Hsml
    ok = unsin->snapshot->getData(comp,"hsml" ,&nbody,&hsml );
    if (ok) 
      unsout->snapshot->setData(comp,"hsml",nbody,hsml,false);  
    // Try to get u
    ok = unsin->snapshot->getData(comp,"u" ,&nbody,&u );
    if (ok) 
      unsout->snapshot->setData(comp,"u",nbody,u,false);  
    // Try to get temperature
    ok = unsin->snapshot->getData(comp,"temp" ,&nbody,&temp );
    if (ok) 
      unsout->snapshot->setData(comp,"temp",nbody,temp,false);  
    // Try to get gas metalicity
    ok = unsin->snapshot->getData(comp,"metal" ,&nbody,&metal );
    if (ok) 
      unsout->snapshot->setData(comp,"gas_metal",nbody,metal,false);  
  }
  if (comp == "stars") {
    float * age, * metal;
    // Try to get stars metalicity
    ok = unsin->snapshot->getData(comp,"metal" ,&nbody,&metal );
    if (ok) 
      unsout->snapshot->setData(comp,"stars_metal",nbody,metal,false);
    // Try to get stars age
    ok = unsin->snapshot->getData(comp,"age" ,&nbody,&age );
    if (ok) 
      unsout->snapshot->setData(comp,"age",nbody,age,false);
  }
  
}
//------------------------------------------------------------------------------
//                             M   A   I   N                                    
//------------------------------------------------------------------------------
int main(int argc, char ** argv )
{
  // global bariable
  std::string typein,file_structin, file_structout;
  //   start  NEMO
  initparam(const_cast<char**>(argv),const_cast<char**>(defv));
  if (argc) {;} // remove compiler warning :)
  
  // Get input parameters
  std::string simname   = (getparam ((char *) "in"      ));
  std::string outname   = (getparam ((char *) "out"     ));
  std::string typeout   = (getparam ((char *) "type"    ));
  std::string select    = (getparam ((char *) "select"  ));
  std::string bits      = (getparam ((char *) "bits"    ));
  std::string select_t  = (getparam ((char *) "times"   ));
  bool        first     = (getbparam((char *) "first"   ));
  float       offset    = (getdparam((char *) "offset"  ));
  bool        verbose   = (getbparam((char *) "verbose" ));
  
  bool one_file=false;
  bool stop=false;
  bool special_nemo=false;
  if (outname=="-" || outname==".") {
    if (typeout!="nemo") {
      std::cerr << "You cannot request out=["<<outname<<"] with type=["<<typeout<<"\n";
      std::exit(1);
    }
    special_nemo=true;
  }
  // in case of an input simulation from the database
  // and with just one time requested,
  // we create a range of time to speedup the searching
  if (select_t!="all" && select_t.find(":",0)==std::string::npos) {
    float match_time;
    stringstream ss;
    ss << select_t;
    ss >> match_time; // convert string time to float
    ss.str(std::string()); // empty stringstream
    ss.clear();            // empty stringstream (mandatory after >>)
    ss << match_time-offset<<":"<<match_time+offset;
    select_t = ss.str();
    one_file=true;
    std::cerr << "Modified selected time =["<<select_t<<"]\n";
  }
  uns::CunsOut * unsout=NULL; // out object
  bool first_out=true;
  
  std::string select_c = select; // default component selection
  if (select.find("@",0) != std::string::npos) { // "@" found
    std::cerr << "Found [@] character in select string\n";
    // we assume that structure file is in=range and out=component
    select_c = "all";
  }
  // -----------------------------------------------
  // instantiate a new UNS input object (for reading)
  uns::CunsIn * unsin = new uns::CunsIn(simname,select_c,select_t,verbose);
  if (unsin->isValid()) { // input file is known by UNS lib        
    int cpt=0;
    while(unsin->snapshot->nextFrame(bits)&&!stop) { // there is a new frame
      typein = unsin->snapshot->getInterfaceType();
      file_structin = unsin->snapshot->getFileStructure();
      std::cerr << "Input file type      :"<<typein<<"\n";
      std::cerr << "Input file structure :"<<file_structin<<"\n";

      int nbody;      
      float time;
      // get the input number of bodies according to the selection
      unsin->snapshot->getData("nsel",&nbody);
      // get the simulation time
      unsin->snapshot->getData("time",&time);
      //      
      std::cerr << "nbody=" << nbody << " time="<<time <<"\n";
      if (nbody>0) { // there are particles
        // OUTPUT operations
        // create an output filename : basename +  integer
        // example : myoutput.0 myoutput.1 ...... etc
        stringstream number;
        number << cpt++;
        std::string out_name=std::string(outname);;
        if (! special_nemo) { // ! standard output && ! "."
          if (one_file || (cpt==1 && !first)) {
            out_name=std::string(outname);
            if (one_file) stop = true; // do not continue
          } else {
            stringstream ss;
            ss << std::string(outname) << "." << setw(5) << setfill('0') << number.str();
            //out_name=std::string(outname)+"."+number.str();
            out_name=ss.str();
          }
          // create a new UNS out object
          unsout = new uns::CunsOut(out_name,typeout,verbose);      
          file_structout = unsout->snapshot->getFileStructure();
        } else {
          if (first_out) {
            first_out = false;
            // instantiate only once unsout, because outname="-"
            unsout = new uns::CunsOut(out_name,typeout,verbose);
            file_structout = unsout->snapshot->getFileStructure();            
          }
        }
        std::cerr << "output filename=["<<out_name<<"]\n";
        std::cerr << "Input  file structure  ="<< file_structin <<"\n";
        std::cerr << "Output file structure  ="<< file_structout <<"\n";
        // save time
        unsout->snapshot->setData("time",time);
        // processing
        if (file_structout=="range") {
          // Whatever input file structure but output file structure is RANGE
          // range= field must be up
          processFromComp("all",unsin,unsout); // only all particles selected
        } else {
          if (file_structin == "range"      && 
              file_structout== "component") {
            std::cerr << "processFromRangeComp\n";
            processFromRangeComp(select,unsin,unsout);
          } else {
            if (file_structin == "component" && 
                file_structout== "component") {
              std::cerr << "processFromComp\n";
              processFromComp("gas"  ,unsin,unsout);
              processFromComp("disk" ,unsin,unsout);
              processFromComp("stars",unsin,unsout);
              processFromComp("bulge",unsin,unsout);
              processFromComp("halo" ,unsin,unsout);
              processFromComp("bndry",unsin,unsout);
            }
          }
        }
        // save snapshot
        unsout->snapshot->save();
        
        if (!special_nemo) {
          delete unsout; // remove object      
        }
      }
    }
  } else {
    std::cerr << "Unknown UNS file format["<<simname<<"]\n";
  }
  delete unsin;
  //   finish NEMO
  finiparam();
}
// ----------- End Of [uns2uns.cc] ------------------------------------
